src/dict.c

```diff
@@ -312,32 +278,107 @@ int dictAdd(dict *d, void *key, void *val)
  *
  * If key was added, the hash entry is returned to be manipulated by the caller.
  */
-dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing)
-{
-    long index;
-    dictEntry *entry;
-    dictht *ht;
-
-    if (dictIsRehashing(d)) _dictRehashStep(d);
-
-    /* Get the index of the new element, or -1 if
-     * the element already exists. */
-    if ((index = _dictKeyIndex(d, key, dictHashKey(d,key), existing)) == -1)
-        return NULL;
-
-    /* Allocate the memory and store the new entry.
-     * Insert the element in top, with the assumption that in a database
-     * system it is more likely that recently added entries are accessed
-     * more frequently. */
-    ht = dictIsRehashing(d) ? &d->ht[1] : &d->ht[0];
-    entry = zmalloc(sizeof(*entry));
-    entry->next = ht->table[index];
-    ht->table[index] = entry;
-    ht->used++;
-
-    /* Set the hash entry fields. */
-    dictSetKey(d, entry, key);
-    return entry;
+dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing) {
+    if (existing) *existing = NULL;
+    if (d->size == 0) {
+        dictSetKey(d, &d->root.entry, key);
+        d->size = 1;
+        return &d->root.entry;
+    }
+    if (d->size == 1) {
```

Comment:
Whether we can make the root always be a `subnode`, so that we can remove the special case `d->size == 1`?

Reply:
I guess it's possible, but then we need some other special cases in other parts of the code, I think...

Reply:
I feel special cases will be less. Maybe I'm wrong.

Reply:
Thanks for the suggestion. I will look into it, but first I want to make the build pass. :-)

Reply:
This PR is difficult. It takes me some time to understand it.